File: test/e2e/clients/php/runner.php
<?php
/**
 * PHP BigQuery client conformance runner.
 *
 * Reads the shared corpus (CASES_FILE), runs every query against the emulator
 * at EMULATOR_HOST with the official google/cloud-bigquery client, then
 * normalizes and compares each result against the case's `expected` block.
 *
 * The comparison logic is intentionally PHP-specific (Date/Timestamp/Bytes
 * wrapper objects) -- verifying that per-language mapping is the point.
 */

require '/deps/php/vendor/autoload.php';

use Google\Cloud\Core\AnonymousCredentials;
use Google\Cloud\BigQuery\BigQueryClient;
use Google\Cloud\BigQuery\Bytes;
use Google\Cloud\BigQuery\Date;
use Google\Cloud\BigQuery\Time;
use Google\Cloud\BigQuery\Timestamp;

function normalize($v)
{
    if ($v === null) {
        return null;
    }
    if (is_bool($v) || is_int($v) || is_float($v) || is_string($v)) {
        return $v;
    }
    if ($v instanceof Date) {
        return $v->formatAsString();
    }
    if ($v instanceof Time) {
        return $v->formatAsString();
    }
    if ($v instanceof Timestamp) {
        return $v->get()->setTimezone(new DateTimeZone('UTC'))->format('Y-m-d\TH:i:s\Z');
    }
    if ($v instanceof Bytes) {
        return base64_encode((string) $v);
    }
    if ($v instanceof DateTimeInterface) {
        return $v->setTimezone(new DateTimeZone('UTC'))->format('Y-m-d\TH:i:s\Z');
    }
    if (is_array($v)) {
        if (array_is_list($v)) {
            return array_map('normalize', $v);
        }
        $out = [];
        foreach ($v as $k => $vv) {
            $out[$k] = normalize($vv);
        }
        return $out;
    }
    return (string) $v;
}

function numbers_equal($a, $b)
{
    return abs((float) $a - (float) $b) <= 1e-9;
}

function values_equal($actual, $expected)
{
    if (is_bool($expected) || is_bool($actual)) {
        return $actual === $expected;
    }
    if (is_numeric($expected) && is_numeric($actual) && !is_string($expected) && !is_string($actual)) {
        return numbers_equal($actual, $expected);
    }
    if (is_array($expected) && is_array($actual)) {
        if (count($expected) !== count($actual)) {
            return false;
        }
        foreach ($expected as $k => $v) {
            if (!array_key_exists($k, $actual) || !values_equal($actual[$k], $v)) {
                return false;
            }
        }
        return true;
    }
    return $actual === $expected;
}

function run_case(BigQueryClient $bq, array $case): array
{
    $config = $bq->query($case['sql']);
    $params = [];
    foreach ($case['params'] ?? [] as $p) {
        $params[$p['name']] = $p['value'];
    }
    if ($params) {
        $config = $config->parameters($params);
    }
    $results = $bq->runQuery($config);

    $rows = [];
    foreach ($results->rows() as $row) {
        $rows[] = normalize($row);
    }
    $expected = $case['expected'];
    if (values_equal($rows, $expected)) {
        return ['name' => $case['name'], 'status' => 'pass', 'detail' => ''];
    }
    return [
        'name' => $case['name'],
        'status' => 'fail',
        'detail' => 'expected ' . json_encode($expected) . ' got ' . json_encode($rows),
    ];
}

$host = getenv('EMULATOR_HOST');
$project = getenv('PROJECT_ID') ?: 'test';
$cases = json_decode(file_get_contents(getenv('CASES_FILE')), true)['cases'];
$reportFile = getenv('REPORT_FILE');

$results = [];
try {
    $bq = new BigQueryClient([
        'projectId' => $project,
        'apiEndpoint' => $host,
        'credentialsFetcher' => new AnonymousCredentials(),
    ]);
    foreach ($cases as $case) {
        try {
            $results[] = run_case($bq, $case);
        } catch (Throwable $e) {
            $results[] = [
                'name' => $case['name'],
                'status' => 'error',
                'detail' => get_class($e) . ': ' . $e->getMessage(),
            ];
        }
    }
} catch (Throwable $e) {
    $detail = get_class($e) . ': ' . $e->getMessage();
    $results = array_map(
        fn ($case) => ['name' => $case['name'], 'status' => 'error', 'detail' => $detail],
        $cases
    );
}

file_put_contents($reportFile, json_encode(['language' => 'php', 'cases' => $results]));
$failed = count(array_filter($results, fn ($r) => $r['status'] !== 'pass'));
echo 'php: ' . count($results) . " case(s), $failed not passing\n";
exit(0);

File: test/e2e/clients/python/runner.py
"""Python BigQuery client conformance runner.

Reads the shared corpus (CASES_FILE), runs every query against the emulator
at EMULATOR_HOST with the official ``google-cloud-bigquery`` client, then
normalizes and compares each result against the case's ``expected`` block.

The comparison logic here is intentionally Python-specific: every client
library returns BigQuery values as different native types, and exercising
that per-language mapping is the whole point of the conformance suite.
"""

import base64
import datetime
import decimal
import json
import os
import sys
import traceback

import requests
from google.api_core import exceptions as api_exceptions
from google.api_core.client_options import ClientOptions
from google.auth.credentials import AnonymousCredentials
from google.cloud import bigquery

# A wedged or unreachable emulator surfaces as one of these; the rest of the
# run is abandoned rather than retried for minutes per remaining case.
TRANSPORT_ERRORS = (
    api_exceptions.RetryError,
    api_exceptions.ServiceUnavailable,
    api_exceptions.DeadlineExceeded,
    requests.exceptions.RequestException,
    ConnectionError,
    TimeoutError,
)

# Per-request and per-job-completion timeout, in seconds.
REQUEST_TIMEOUT = 20


def normalize(value):
    """Map a value returned by the Python client to the canonical form."""
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        return value
    if isinstance(value, decimal.Decimal):
        return str(value)
    if isinstance(value, (bytes, bytearray)):
        return base64.b64encode(bytes(value)).decode("ascii")
    if isinstance(value, datetime.datetime):
        utc = value.astimezone(datetime.timezone.utc) if value.tzinfo else value
        return utc.strftime("%Y-%m-%dT%H:%M:%SZ")
    if isinstance(value, datetime.date):
        return value.isoformat()
    if isinstance(value, datetime.time):
        return value.strftime("%H:%M:%S")
    if isinstance(value, (list, tuple)):
        return [normalize(v) for v in value]
    if isinstance(value, dict):
        return {k: normalize(v) for k, v in value.items()}
    if hasattr(value, "items"):  # bigquery.Row exposes items()
        return {k: normalize(v) for k, v in value.items()}
    return str(value)


def numbers_equal(a, b):
    return abs(float(a) - float(b)) <= 1e-9


def values_equal(actual, expected):
    if isinstance(expected, bool) or isinstance(actual, bool):
        return actual == expected
    if isinstance(expected, (int, float)) and isinstance(actual, (int, float)):
        return numbers_equal(actual, expected)
    if isinstance(expected, list) and isinstance(actual, list):
        return len(actual) == len(expected) and all(
            values_equal(x, y) for x, y in zip(actual, expected)
        )
    if isinstance(expected, dict) and isinstance(actual, dict):
        return set(actual) == set(expected) and all(
            values_equal(actual[k], expected[k]) for k in expected
        )
    return actual == expected


def make_param(spec):
    return bigquery.ScalarQueryParameter(spec["name"], spec["type"], spec["value"])


def run_case(client, case):
    params = [make_param(p) for p in case.get("params", [])]
    job_config = bigquery.QueryJobConfig(query_parameters=params)
    # Bound both the per-request timeout and the overall retry window so a
    # wedged emulator fails the case quickly instead of retrying for minutes.
    retry = bigquery.DEFAULT_RETRY.with_timeout(REQUEST_TIMEOUT)
    job = client.query(
        case["sql"], job_config=job_config, timeout=REQUEST_TIMEOUT, retry=retry
    )
    rows = [normalize(dict(row)) for row in job.result(timeout=REQUEST_TIMEOUT, retry=retry)]
    expected = case["expected"]
    if values_equal(rows, expected):
        return {"name": case["name"], "status": "pass", "detail": ""}
    return {
        "name": case["name"],
        "status": "fail",
        "detail": "expected {} got {}".format(
            json.dumps(expected), json.dumps(rows)
        ),
    }


def main():
    host = os.environ["EMULATOR_HOST"]
    project = os.environ.get("PROJECT_ID", "test")
    cases_file = os.environ["CASES_FILE"]
    report_file = os.environ["REPORT_FILE"]

    with open(cases_file) as fh:
        cases = json.load(fh)["cases"]

    results = []
    try:
        client = bigquery.Client(
            project,
            client_options=ClientOptions(api_endpoint=host),
            credentials=AnonymousCredentials(),
        )
        aborted = None
        for case in cases:
            if aborted is not None:
                results.append({"name": case["name"], "status": "error", "detail": aborted})
                continue
            try:
                results.append(run_case(client, case))
            except TRANSPORT_ERRORS as exc:
                # The emulator is unreachable or unresponsive; abandon the rest
                # of the run instead of retrying every remaining case.
                aborted = "aborted: emulator unreachable/unresponsive: {}".format(exc)
                results.append({"name": case["name"], "status": "error", "detail": aborted})
            except Exception:  # noqa: BLE001 - one bad case must not abort the run
                results.append(
                    {
                        "name": case["name"],
                        "status": "error",
                        "detail": traceback.format_exc(limit=3),
                    }
                )
    except Exception:  # noqa: BLE001 - client setup failed; mark every case
        detail = traceback.format_exc(limit=3)
        results = [
            {"name": c["name"], "status": "error", "detail": detail} for c in cases
        ]

    report = {"language": "python", "cases": results}
    with open(report_file, "w") as fh:
        json.dump(report, fh)

    failed = sum(1 for r in results if r["status"] != "pass")
    print("python: {} case(s), {} not passing".format(len(results), failed))
    sys.exit(0)


if __name__ == "__main__":
    main()
